### PR TITLE
--story=IGL | opengl | support glDrawElementsInstanced

### PR DESCRIPTION
**Pull request by vinsentli**

### CODE REVIEW

File: src/igl/opengl/IContext.cpp [hunk 1]
```diff
@@ -1541,6 +1541,23 @@ void IContext::drawElements(GLenum mode, GLsizei count, GLenum type, const GLvoi
   APILOG_DEC_DRAW_COUNT();
 }
 
+void IContext::drawElementsInstanced(GLenum mode, GLsizei count, GLenum type, const GLvoid* indices, GLsizei instancecount) {
+    drawCallCount_++;
+
+    IGL_PROFILER_ZONE_GPU_COLOR_OGL("drawElementsInstanced()", IGL_PROFILER_COLOR_DRAW);
+
+    GLCALL(DrawElementsInstanced)(mode, count, type, indices, instancecount);
```
Comment (corporateshark) [inline src/igl/opengl/IContext.cpp:1549]: This is not available on GLES 2 and GLES 3.0.

File: src/igl/opengl/RenderCommandEncoder.cpp [hunk 2]
```diff
@@ -398,8 +396,11 @@ void RenderCommandEncoder::drawIndexed(PrimitiveType primitiveType,
   if (IGL_VERIFY(adapter_ && indexType_)) {
     getCommandBuffer().incrementCurrentDrawCount();
     auto mode = toGlPrimitive(primitiveType);
-    adapter_->drawElements(
-        mode, (GLsizei)indexCount, indexType_, (uint8_t*)indexBufferOffset_ + indexOffsetBytes);
+    if (instanceCount > 1){
```
Comment (corporateshark) [inline src/igl/opengl/RenderCommandEncoder.cpp:399]: Should be checked against currently available OpenGL or OpenGL ES version.

File: src/igl/opengl/RenderPipelineState.cpp [hunk 3]
```diff
@@ -303,7 +303,13 @@ void RenderPipelineState::bindVertexAttributes(size_t bufferIndex, size_t buffer
         reinterpret_cast<const char*>(attribute.bufferOffset) + bufferOffset);
 
     if (getContext().deviceFeatures().hasInternalFeature(InternalFeatures::VertexAttribDivisor)) {
-      getContext().vertexAttribDivisor(location, 0);
+        if (attribute.sampleFunction == igl::VertexSampleFunction::PerVertex){
```
Comment (corporateshark) [inline src/igl/opengl/RenderPipelineState.cpp:306]: This should be done in a separate PR.

**Review by corporateshark — CHANGES_REQUESTED** [review]

File: src/igl/opengl/IContext.cpp (hunk 1 above)
Comment (vinsentli) [reply src/igl/opengl/IContext.cpp:1549]: This is available on GLES 3.0, not available on GLES 2. I will improve it.

Comment (corporateshark) [reply src/igl/opengl/IContext.cpp]: Sorry, not available in desktop OpenGL 3. Not GLES 3. :-)

https://registry.khronos.org/OpenGL-Refpages/gl4/html/glDrawElementsInstanced.xhtml

**Review by corporateshark — APPROVED** [review]

**Comment by vinsentli** [comment]
@corporateshark 
There is also glDrawArraysInstanced that needs to be implemented. I will add another pull request for that.

**Review by corporateshark — CHANGES_REQUESTED** [review]
@vinsentli Please keep the changes adding `glDrawElementsInstances()` and `glDrawArraysInstanced()` in two separate PRs. It will help us to test and review the new code much faster.

**Comment by vinsentli** [comment]
> @vinsentli Please keep the changes adding `glDrawElementsInstances()` and `glDrawArraysInstanced()` in two separate PRs. It will help us to test and review the new code much faster.

OK , I revert the commit about glDrawArraysInstanced().
I will pull another request about glDrawArraysInstanced()  after glDrawElementsInstances() approved. 
If I pull two requests at the same time may cause merge conflict.

**Comment by vinsentli** [comment]
@corporateshark 
This PR is about glDrawElementsInstances().
There are two PRs on the way,  glDrawArraysInstanced() and glDrawArraysIndirect().

**Comment by vinsentli** [comment]
Resolve conflicts.
